### PR TITLE
bpo-41818: Updated tests for the standard pty library

### PR DESCRIPTION
**Pull request by 8vasu**
This converts the test for pty.master_open() to a test for pty.openpty() with additional tests such as checking slave termios, slave winsize; these additional tests are expected to fail.

This also adds a small test that is expected to fail on FreeBSD and should pass on Linux; reflecting pty.spawn()'s FreeBSD hang issue [ [bpo-26228](https://bugs.python.org/issue26228) ].

Solutions to these problems have been implemented here: https://github.com/8vasu/pypty2

If and when these tests are acknowledged, the solutions will be submitted in parts.

This follows the conversation here: https://github.com/python/cpython/pull/21861

Signed-off-by: Soumendra Ganguly <soumendraganguly@gmail.com>

https://bugs.python.org/issue41818

### CODE REVIEW

**Comment by 8vasu** [comment]
@ethanfurman @aeros

Sir, thank you very much for being patient. Please run the tests when you have time. I welcome suggestions/criticism. I have also sent a message to the teaching email thread.

File: Lib/test/test_pty.py
Comment (gpshead) [inline Lib/test/test_pty.py]: please get rid of this leftover commented out pty2 import :)

**Review by gpshead — COMMENTED** [review]
i kicked off a test on the buildbots for some additional platform coverage out out curiosity.  **Note**: our buildbot fleet may generate a lot of noise, they aren't all reliably and some were already failing on the main branch to begin with right now.  so don't fret if it claims there are failures.  only look for obviously related things within them if you go looking at those. :)

**Comment by 8vasu** [comment]
@gpshead I have removed those comments :)

File: Lib/test/test_pty.py
Comment (aeros) [inline Lib/test/test_pty.py:88]: `unittest.skipTest()` is typically used for unsupported platforms in tests, rather than `expectedFailure()`. The latter tends to imply that the test is specifically testing a behavior that's expected to not work.

File: Lib/test/test_pty.py
Comment (aeros) [inline Lib/test/test_pty.py]: Can we trim down the debug messages a bit here? To me, explicitly having "Calling pty.openpty()", "Closing slave_fd", and "Reading from master_fd" doesn't provide much valuable additional context to someone trying to resolve a failure.

Also, for "Got master_fd '%d', slave_fd '%d'", it would be preferable to use f-strings for better readability, with it being an entirely new test (although I personally think it's fine to leave `test_openpty()` as-is with using % interpolation, since that's what it used previously).

File: Lib/test/test_pty.py
Comment (aeros) [inline Lib/test/test_pty.py:364]: It's not clear to me why a new timeout parameter was added here, could you clarify?

File: Lib/test/test_pty.py
Comment (aeros) [inline Lib/test/test_pty.py]: Shouldn't this at least result in a warning via `warnings.warn()`? I suspect it would get buried otherwise.

File: Lib/test/test_pty.py
Comment (aeros) [inline Lib/test/test_pty.py]: This existing whitespace line should be re-added, since generally there are two lines in between the last line of a class's body and functions after it.

File: Misc/NEWS.d/next/Library/2020-10-26-14-17-44.bpo-41818.5MgaEc.rst
Comment (aeros) [inline Misc/NEWS.d/next/Library/2020-10-26-14-17-44.bpo-41818.5MgaEc.rst]: The news entry should optimally refer to the tests that were changed (in this case `test_opentty` was updated and a new `test_master_read()` was added.

File: Lib/test/test_pty.py
Comment (aeros) [inline Lib/test/test_pty.py]: I don't think this comment adds any additional context for the reader, and can be removed.

File: Lib/test/test_pty.py
Comment (aeros) [inline Lib/test/test_pty.py]: Maybe @ethanfurman or @Yhg1s could clarify, but I don't think we should add a new "RELEVANT ANYMORE?" comment.

File: Lib/test/test_pty.py
Comment (aeros) [inline Lib/test/test_pty.py]: Could we potentially surround with a `try...finally` to avoid this? A single test failure should really not result in the other tests within `test_pty` being unreliable.

**Review by aeros — COMMENTED** [review]
I've completed a preliminary review of the test changes. Most of the suggested changes are fairly minor, for the most part I think this is good to go (although for the final approval/merge, I'd prefer to leave it to @ethanfurman or @Yhg1s since I'm no expert with `pty` or `tty`).

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py:88]: Actually, in this case, the purpose IS to demonstrate to the reviewer that the test will fail on the BSDs. This test pinpoints the root cause of `pty.spawn()`'s infamous FreeBSD [ actually *BSD + macOS + ... ] hang issue. Basically, once 1. the child process stops AND 2. ALL descriptors of the `slave` are closed by the parent, an OSError is raised on Linux when the parent attempts to read from the `master`; the BSDs on the other hand simply return EOF; the current implementation "wrongly" depends on this Linux idiosyncrasy to exit from the parent main loop. Further, it advertises this "bug" as a feature: https://docs.python.org/3/library/pty.html says "in order to force spawn to return before the child process exits an OSError should be thrown". Once the fix for this is added [ which is in pypty2 as I have mentioned before ], *BSDs, macOS, etc will become supported platforms. Then we can safely comment out or remove this test. Also, if the strange "throw OSError to exit from loop" feature needs to be retained for backward compatibility, all I would need to do is add one line to pypty2.

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py]: Good point. Thank you. I will work on this.

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py]: You are absolutely correct. I will change this. I have one question before I change this. If I use warnings.warn() here, then it will always generate warnings when the test is not run with stdin being a tty; for example the automated tests done by GitHub. Is that going to be an issue?

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py]: Thank you for noticing this. I agree that they are annoying; I apologize for this. Yes, I need clarification. The plan was to remove the annoying "RELEVANT ANYMORE?" comments after clarification.

Comment (8vasu) [reply Lib/test/test_pty.py]: I myself actually did not put much thought into the blocks that I marked with "RELEVANT ANYMORE?"; I will think about them in greater detail on Saturday. They were just supposed to be there as reminders; anyway, they do not hurt the tests :)

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py]: Gotcha. I did this because the test_fork() function also has a similar comment at the end thereof. Should I remove that one as well?

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py]: Sure. I retained them because that's what the older function was doing.  I support the idea of using f-strings here. I will make these changes.

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py]: Thank you for teaching me this. I will make this change.

File: Misc/NEWS.d/next/Library/2020-10-26-14-17-44.bpo-41818.5MgaEc.rst
Comment (8vasu) [reply Misc/NEWS.d/next/Library/2020-10-26-14-17-44.bpo-41818.5MgaEc.rst]: Sure!

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py:364]: Yes! Glad that you asked! This is actually an interesting story. `SmallPtyTests.test__copy_eof_on_all()` emulates a situation where both `stdin` and `master` reach EOF in the same iteration of `pty.spawn()`'s copy loop ( `_copy()` in Lib/pty.py ). This leads to both of them being removed from `select()`'s fd list at once. While this does not occur often in practical scenarios on Linux, FreeBSD, etc, this is a situation that must be considered [ maybe some other platform behaves this way/some special case ]; that is why this test exists in the first place! Now, at least on Linux, select.select([],[],[]) will block indefinitely [ all three args are empty lists because `master` and `stdin` were kicked out ]. In such situation, one should set a timeout for `select.select()` once both `master` and `stdin` are out. After the timeout is reached, we exit from the loop gracefully. Now, one might ask why not just put a `while fds:` there instead of the `while True` to avoid setting a timeout. The reason is that `SmallPtyTests.test__copy_eof_on_all()` EXPECTS the copy loop to make ONE ADDITIONAL iteration of the loop before it exits. In fact, this is how I have seen it being done in other programs; the util-linux version of script(1) sets a timeout for `poll()` instead of abruptly exiting from its main loop when no more data is available. Therefore, pypty2 also does it this way, so that `SmallPtyTests.test__copy_eof_on_all()` works as expected without modification; only `_mock_select()` must now accept a timeout arg to be compatible; anyway, it ignores the timeout :D whereas the actual `select.select()` does not.

**Comment by 8vasu** [comment]
@aeros I am grateful for the reviews. I will keep working on this with the hope that we can finally put these old bugs to rest.

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py:88]: Further note: actually throwing an OSError in master_read()/stdin_read() will currently only get you out of the copy loop ( `_copy()`); if the child is still running, then `pty.spawn()` will additionally block at the final `os.waitpid()` call.

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py:364]: Update: I am willing to remove the timeout hack in `pty._copy()` and replace it with the cleaner `while fds:` alternative that I mentioned. However, that will require modifying/removing `SmallPtyTests.test__copy_eof_on_all()`, which I should only do after I have convinced the developers that everything else works correctly :)

File: Lib/test/test_pty.py
Comment (aeros) [reply Lib/test/test_pty.py]: It may cause a slight bit of extra noise on the tests, but it shouldn't cause the GitHub CI to fail when the warning is emitted (I don't believe we have any of our CI set to fail on warning, but it can be ran with that setting locally).

File: Lib/test/test_pty.py
Comment (8vasu) [reply Lib/test/test_pty.py]: Aah. Now I realize why I added the extra "RELEVANT ANYMORE" here. Notice that the following comment mentions `pty.master_open()`, which opens the slave and subsequently closes it. As a result, the original `test_basic()` function was having to call `pty.slave_open()` to re-open the slave. However, the new `test_openpty()` calls `pty.openpty()`, which, unlike `pty.master_open()`, does not close the slave end. Anyway, that part does not seem to affect the test on Linux and FreeBSD.

**Comment by 8vasu** [comment]
@aeros  Thank you for being patient. I have made the following changes as per your suggestions:

1. I had added a comment saying that the tty could be left in an abnormal state [ only winsize, not termios/terminfo ]. I have solved the problem by adding an addCleanup(). I have also renamed some variables, added functions [ _get_term_winsz() and _set_term_winsz() ] to increase readability.

2. I have changed the `debug()` in the except clause of the second try-except block in test_openpty() to a `warnings.warn()` as per your suggestion.

3. I realized why I added the extra "RELEVANT ANYMORE" comment. The following comment mentions `pty.master
_open()`, which opens the slave and subsequently closes it. As a result, the original `test_basic()` function was having t
o call `pty.slave_open()` to re-open the slave. However, the new `test_openpty()` calls `pty.openpty()`, which, unlike `pty.ma
ster_open()`, does not close the slave end. Anyway, that part does not seem to affect the test on Linux and FreeBSD.
I have removed the "RELEVANT ANYMORE" comments.

4. Removed the "pty.openpty() passed" and "pty.fork() passed" comments.

5. Replaced % with f-strings. However, I have not trimmed down the debug messages yet because `test_fork()` has similar debug messages in it. I will remove them all at once if you wish.

6. Added the existing empty line between class and subsequent function definition.

File: Lib/test/test_pty.py
Comment (ethanfurman) [inline Lib/test/test_pty.py]: I've been away from the code, so this might be a dumb question, but can't we determine if `stdin` is a tty?

Comment (8vasu) [reply Lib/test/test_pty.py]: Sorry for the late reply. That is actually a great question. `tcgetattr()` can fail due to 2 reasons: `errno == EBADF` ( not a valid file descriptor ), `errno == ENOTTY` ( not a tty ). See https://pubs.opengroup.org/onlinepubs/9699919799/. In either case, your stdin is not a tty. In fact, all glibc's `isatty()` does is call `tcgetattr()` to see if it fails; see https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/posix/isatty.c.

I have verified that Linux, FreeBSD, and NetBSD follow this approach for their `isatty()` implementations, while OpenBSD calls `fcntl()+F_ISATTY`. Anyway, the `tcgetattr()` approach works on all the above platforms. I avoid `isatty()`, which is often followed by a separate `tcgetattr()` call, thereby avoiding an extra `tcgetattr()` call [ OCD :D ].

In our code, if `tcgetattr()` fails, then either `pty.STDIN_FILENO`, which is `0`, is a valid fd but not a tty (ENOTTY), or it is not a valid fd at all (EBADF). EBADF will not happen unless something closes `0` [ uncommon scenario ]; for example, by using [ `p close(0)` ] in `gdb`, or if your script itself voluntarily closes `0` for some reason.

The "possible" in the comment  `# possible reason: current stdin is not a tty` is not because of lack of confidence, but because I did not want to discount the (uncommon) chance of an EBADF, in which case `0` is not a valid fd at all [ OCD ]. Anyway, that comment can be safely removed. Do you want me to remove it?

Comment (ethanfurman) [reply Lib/test/test_pty.py]: I would say update it:

`# possible reason ...` to `# not a tty or bad/closed fd`

and, of course, make that change throughout.

Comment (8vasu) [reply Lib/test/test_pty.py]: Done.

**Review by asvetlov — APPROVED** [review]
LGTM, I think we can merge the PR and unblock @8vasu to work on future PTY improvements.

**Comment by asvetlov** [comment]
The test fails with 'unexpected success' error, perhaps `expectedFailure()` should not be applied for this configuration.
Investigating.

**Comment by ethanfurman** [comment]
Or perhaps Gentoo is one of the cases where it succeeds, and the the expected failure wrapper should not be applied in that case.

**Comment by asvetlov** [comment]
#23514 is created for fix

**Pull request merged**
